### ISSUE TITLE
Data race between multiple duchess versions (in theory)

### ISSUE THREAD

**Issue by fpoli** [duchess-rs/duchess#59]
I think that the race condition fixed in https://github.com/duchess-rs/duchess/pull/53 can still be triggered (in theory) by either
* writing a crate that depends on *multiple versions* of `duchess`, so that each `duchess` gets its own `GLOBAL_JVM`, or
* using `duchess` together with other (sound, on their own) libraries that launch a JVM via JNI.

I'm not sure if there is a way for `duchess` (or any other library) to launch a JVM without risking the data race. JNI being a process-global but not thread-safe library is quite restrictive. In the worst case, it might be enough to just state the risk in the documentation.

**Comment by nikomatsakis** [comment]
That reminds me -- I wanted to write up some sections on using duchess safely, talking about exactly this sort of point. It'd be nice if there was some way to synchronize the JNI launching on the *JNI* side.

**Comment by chase-ok** [comment]
Yeah, this looks difficult to enforce either with combinations of JNI wrappers (`duchess` + `jni` crates) or with combinations of Duchess versions. Given the unsafety inside of `libjvm`, not sure what more we can do.

**Comment by fpoli** [comment]
I did some investigation. It seems that (in OpenJDK) concurrent calls to `JNI_CreateJavaVM` work fine because internally they use an atomic integer to check and update the number of created JVMs ([source](https://github.com/openjdk/jdk/blob/8d49ba9e8d3095f850b3007b56488a0c0cf8ddff/src/hotspot/share/prims/jni.cpp#L3544-L3546)). However, `JNI_GetCreatedJavaVMs` happily returns a pointer to a JVM that is still being constructed by another thread ([source](https://github.com/openjdk/jdk/blob/8d49ba9e8d3095f850b3007b56488a0c0cf8ddff/src/hotspot/share/prims/jni.cpp#L3671)). That's the reason of the crash.

For example, running `race_multiple_threads_to_launch_jvm` on the buggy duchess with some print statements before and after each JNI call can lead to:
```
[ThreadId(1)] BEGIN JNI_CreateJavaVM
[ThreadId(2)] BEGIN JNI_CreateJavaVM
[ThreadId(2)] END JNI_CreateJavaVM returns code -5, jvm 0x0
[ThreadId(2)] BEGIN JNI_GetCreatedJavaVMs
[ThreadId(2)] END JNI_GetCreatedJavaVMs returns code 0, num_jvms 1, jvms [0x7ffff75114a0]
[ThreadId(2)] AttachCurrentThread uses jvm 0x7ffff75114a0
<crash>
```
Note that `[ThreadId(1)] END JNI_CreateJavaVM` is missing.

**Comment by fpoli** [comment]
A note in the documentation sounds enough for me.

**Comment by fpoli** [comment]
> Maybe we can file a bug on the OpenJDK

I already reported a bug. I think it remains private until someone reviews it.

**Comment by fpoli** [comment]
> I already reported a bug. I think it remains private until someone reviews **it.

The bug report is [JDK-8308341](https://bugs.java.com/bugdatabase/view_bug?bug_id=JDK-8308341)

**Comment by nikomatsakis** [comment]
Discussion on Zulip:

https://duchess.zulipchat.com/#narrow/stream/388667-general/topic/co-existing.20with.20other.20jvm.20wrappers/near/363468859
